crates/core/src/rescue.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::{Deref, DerefMut};
+use std::sync::{Arc, Mutex};
+
+pub type Rescued<T> = Arc<Mutex<Option<T>>>;
+
+pub trait Recoverable: Sized {
+    fn rescue(&mut self) -> Option<Self>;
+}
+
+#[derive(Default)]
+pub struct Wrapper<T: Recoverable> {
```

Comment:
An alternative to "abusing" the drop to get the Database back is to move the DB in and out similar to https://github.com/risc0/risc0-ethereum/blob/73eebd4efb283c7f99748d45fc730a53ee2edf5c/steel/src/contract.rs#L211
However, this should probably not also be in this PR.

Reply:
The code to do this would have be on the reth side so to keep the changes there minimal I went with rescuing the dropped value